lib/rules.js

```diff
@@ -0,0 +1,211 @@
+  }
+  return newRule;
+};
+
+class FunctionRule extends Rule {
+  constructor(fn, opts = { modifiesCharacters: true }) {
+    if (!opts.name) opts.name = fn.toString();
+    super(opts);
+    this.fn = fn;
+    this.type = 'FunctionRule';
+  }
+
+  apply(s1, s2) {
+    return this.fn(s1, s2);
+  }
+}
+
+// TODO: ideally, MapRule would mixin Map.
+class MapRule extends Rule {
+  constructor(map, opts = { modifiesCharacters: false }) {
+    if (!opts.name) opts.name = map.toString();
+    super(opts);
+    assert(
+      map instanceof Map,
+      `${map} is not an instance of Map (is ${map.constructor})`
+    );
+    this.map = map;
+    this.type = 'MapRule';
+  }
+
+  apply() {
+    throw new Error('MapRules cannot be directly applied.');
+  }
+
+  entries() {
+    return this.map.entries();
+  }
+
+  keys() {
+    return this.map.keys();
+  }
+}
+
+class RemoveRule extends MapRule {
+  constructor(str, opts = { modifiesCharacters: false }) {
+    assert(
+      typeof str === 'string',
+      `${str} is not a string (is ${str.constructor})`
+    );
+    const map = mapFromString(str);
+    super(map, opts);
+  }
+}
+
+const identity = (s1, s2) => [s1, s2];
+const identityRule = new FunctionRule(identity, { name: 'identity' });
+
+const capitalization = (s1, s2) => {
+  return [s1.toLowerCase(), s2.toLowerCase()];
+};
+const capitalizationRule = new FunctionRule(capitalization, {
+  name: 'capitalization',
+});
+
+const unicodeNormalization = (s1, s2) => {
+  return [normalize(s1, 'NFC'), normalize(s2, 'NFC')];
+};
+const unicodeNormalizationRule = new FunctionRule(unicodeNormalization, {
+  name: 'unicodeNormalization',
+});
+
+const RE_ANY_WHITESPACE = /[\s]+/g;
+const whitespaceDifferences = (s1, s2) => {
+  return [
+    s1.trim().replace(RE_ANY_WHITESPACE, ' '),
+    s2.trim().replace(RE_ANY_WHITESPACE, ' '),
+  ];
+};
+const whitespaceDifferencesRule = new FunctionRule(whitespaceDifferences, {
+  name: 'whitespaceDifferences',
+});
+
+const acuteAccentCombining = '\u0301';
+// Note that Unicode calls this "diaresis", see
+// https://en.wikipedia.org/wiki/Diaeresis_(diacritic)
+const umlautCombining = '\u0308';
+
+// https://www.fileformat.info/info/unicode/block/combining_diacritical_marks/list.htm
+// U+0300–U+036F
+const startOfCombiningDiacriticsBlock = 0x300;
+const endOfCombiningDiacriticsBlock = 0x36f;
+const combiningDiacriticsBlock = new Array(
+  endOfCombiningDiacriticsBlock - startOfCombiningDiacriticsBlock
+)
+  .fill(undefined)
+  .map((_, idx) => String.fromCodePoint(idx + startOfCombiningDiacriticsBlock));
+
+const commonCombiningDiacritics = [
+  '\u0300', // grave accent
+  acuteAccentCombining, // acute accent
+  '\u0302', // circumflex
+  '\u0303', // tilde
+  umlautCombining, // diaresis, aka umlaut
+  '\u0327', // cedilla
+];
+
+// pass in an array of strings
+const codepointRuleGenerator = (codepoints, name) => {
+  return new FunctionRule(
+    (s1, s2) => {
+      let s1prime = normalize(s1, 'NFD');
```

Comment:
This produces Unicode denormalized strings, where each combining character is its own character (so there's a character for an accent separate from the character it appears above, for example). Then we can remove all the diacritics from both sides, making them effectively not matter. The trick with determining whether this was the reason for two strings not to be equivalent is if they're equivalent when this rule doesn't matter, but not equivalent when it does, then this rule is responsible for them differing (that part of the code is above in index.js).